explauto/interest_model/riac.py

```diff
@@ -0,0 +1,947 @@
+        
+        Raises
+        ------
+        RuntimeError
+            The maximum recursion limit can be exceeded for large data
+            sets.  If this happens, either increase the value for the `max_points_per_region`
+            parameter or increase the recursion limit by::
+    
+                >>> import sys
+                >>> sys.setrecursionlimit(10000)
+    
+
+    """
+    def __init__(self, 
+                 get_data_x, 
+                 bounds_x, 
+                 get_data_c, 
+                 max_points_per_region=10, 
+                 split_mode='median', 
+                 progress_win_size=5, 
+                 progress_measure='abs_deriv', 
+                 sampling_mode=['softmax',0.1], 
+                 idxs=[], 
+                 split_dim=0):
+        
+        self.get_data_x = get_data_x
+        self.bounds_x = np.array(bounds_x, dtype=np.float64)
+        self.get_data_c = get_data_c
+        self.max_points_per_region = max_points_per_region
+        self.split_mode = split_mode
+        self.progress_win_size = progress_win_size
+        self.progress_measure = progress_measure
+        self.sampling_mode = sampling_mode
+
+        self.split_dim = split_dim
+        self.split_value = None
+        self.less = None
+        self.greater = None
+        self.idxs = idxs
+        self.children = len(self.idxs)
+        
+        self.leafnode = True
+        self.progress = 0
+        
+        
+        if self.children > self.max_points_per_region:
+            self.split()
+        self.compute_progress()
+            
+        
+    def get_leaves(self):
+        """
+        Get the list of all leaves
+        
+        """
+        return self.fold_up(lambda fl, fg: fl + fg, lambda leaf: [leaf])
+    
+    
+    def depth(self):
+        """
+        Compute the depth of the tree (depth of a leaf=0)
+        
+        """
+        return self.fold_up(lambda fl, fg: max(fl + 1, fg + 1), lambda leaf: 0)
+    
+    
+    def volume(self):
+        """
+        Compute the volume of the node
+        
+        """
+        return np.prod(self.bounds_x[1,:] - self.bounds_x[0,:])
+    
+    
+    def density(self):
+        """
+        Compute the density of the node
+        
+        """
+        return self.children / self.volume()
+    
+    
+    def pt2leaf(self, x):
+        """
+        Get the leaf which domain contains x
+        
+        """
+        if self.leafnode:
+            return self
+        else:
+            if x[self.split_dim] < self.split_value:
+                return self.less.pt2leaf(x)
+            else:
+                return self.greater.pt2leaf(x)
+        
+        
+    def sample_bounds(self):
+        """
+        Sample a point in the region of this node
+        
+        """
+        s = rand_bounds(self.bounds_x).flatten()
+        return s
+    
+        
+    def sample(self, sampling_mode = None):
+        """
```

Comment:
As you have quite a lot of sampling_mode it would be nice to gather them in some sort of struct (or at least to have the possibilities in the docstring).

Maybe using an [enum](https://docs.python.org/3/library/enum.html)? Maybe an overkill here?